### ISSUE TITLE
 modelutils/src/utils/fst_read.F90 : 32-bits integer is not sufficient for variable dist and dist0

### ISSUE THREAD

**Issue by frank2243** [armnlib/modelutils#1]
At each SST (TM and LG) timestep, the model read variables at first in the analysis file (initial condition of the run) and then in the driving data. However, during that process the subroutine fst_read.F90 calculate the time difference in second between the analysis file date and the actual date of the simulation. This could cause problem for long run (~60 years) as the difference in seconds could become to large for a 32-bits integer. 

As a fix to this the following changes are proposed : 

The following two new variables have to be declared at the beginning of the file : 
`integer*8 :: dist64, dist064`

Here are the original problematic lines (171 to 200) : 
```
  dist = datevfuzz
      if (datevfuzz /= huge(datevfuzz)) dist = datevfuzz+min_time_res
      do i=1,nkeys
         istat = fstprm(keylist(i), dateo,deet,npas, ni1,nj1,nk1, &
              nbits, datyp, ip1, ip2, ip3, &
              typvar1_S, nomvar_S, etiket_S, &
              grtyp_S, ig1, ig2, ig3, ig4, swa, lng, dltf, &
              ubc, extra1, extra2, extra3)
         if (.not.RMN_IS_OK(istat)) then
            F_key = RMN_ERR
            call msg(MSG_WARNING,'(fst_find) problem getting record meta')
            return
         endif
         nhours_8 = (DBLE(deet)*DBLE(npas))/SEC_PER_HR
         call incdatr(datev0,dateo,nhours_8)
         call difdatr(searchdate,datev0,nhours_8)
         dist0 = nint(real(nhours_8*SEC_PER_HR))
         select case(fuzzopr)
         case(FST_FIND_GE)
            dist0 = -dist0
         case(FST_FIND_NEAR)
            dist0 = abs(dist0)
         !case(FST_FIND_LE)
         end select
         if (dist0 >= 0 .and. dist0 < dist) then
            F_key = keylist(i)
            dist  = dist0
            datev = datev0
         endif
      end do
```

Here are the changes lines : 
```
      dist64 = int(datevfuzz,kind=8)
      if (datevfuzz /= huge(datevfuzz)) dist64 = int(datevfuzz+min_time_res,kind=8)
      do i=1,nkeys
         istat = fstprm(keylist(i), dateo,deet,npas, ni1,nj1,nk1, &
              nbits, datyp, ip1, ip2, ip3, &
              typvar1_S, nomvar_S, etiket_S, &
              grtyp_S, ig1, ig2, ig3, ig4, swa, lng, dltf, &
              ubc, extra1, extra2, extra3)
         if (.not.RMN_IS_OK(istat)) then
            F_key = RMN_ERR
            call msg(MSG_WARNING,'(fst_find) problem getting record meta')
            return
         endif
         nhours_8 = (DBLE(deet)*DBLE(npas))/SEC_PER_HR
         call incdatr(datev0,dateo,nhours_8)
         call difdatr(searchdate,datev0,nhours_8)
         
         dist064 = nint(real(nhours_8*SEC_PER_HR,kind=8),kind=8)

         select case(fuzzopr)
         case(FST_FIND_GE)
	    dist064 = -dist064
         case(FST_FIND_NEAR)
	    dist064 = abs(dist064)
         !case(FST_FIND_LE)
         end select
	 if (dist064 >= 0 .and. dist064 < dist64) then
            F_key = keylist(i)
            dist64  = dist064
            datev = datev0
         endif
      end do
```

Basically, the variables dist and dist0 were switched to 64-bits integers and the appropriate explicit type conversion were added to make it more clear to the reader what is converted and why.

**Comment by LaurentChardon** [comment]
Thanks for the heads up. We'll have a look into this.

**Comment by frank2243** [comment]
Merci.
Thank you.

**Comment by jeixav** [comment]
This issue was forwarded to the CMC intranet bugzilla: http://bugzilla.cmc.ec.gc.ca/show_bug.cgi?id=9680.
